### PR TITLE
Add ECS relevant info for container insight receiver

### PR DESCRIPTION
**Pull request by JohnWu20**
**Description:** 
This commit add the ecs info that are used by container insight receiver for ecs instance level metrics. This includes:

Get task info, clusterName and containerInstanceId from Amazon [ECS container agent introspection](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-introspection.html).

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2307

**Testing:** <Describe what testing was performed and which tests were added.>
Unit test

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by alolita** [comment]
Hi @JohnWu20 please fix the failing tests listed below. 

    windows-test - Failed
    integration-tests - Failed
    unit-tests - Failed
    lint - Failed

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+)
+
+type HttpClient struct {
+	maxRetries              int
+	backoffRetryBaseInMills int
+	client                  clientProvider
+}
+
+type HttpClientProvider interface {
+	Request(path string, ctx context.Context, logger *zap.Logger) ([]byte, error)
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:44]:
```suggestion
	Request(ctx context.Context, path string, logger *zap.Logger) ([]byte, error)
```

`ctx` should always be the first argument.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	maxHttpResponseLength          = 5 * 1024 * 1024 // 5MB
+)
+
+type HttpClient struct {
+	maxRetries              int
+	backoffRetryBaseInMills int
+	client                  clientProvider
+}
+
+type HttpClientProvider interface {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:43]:
```suggestion
type Requester interface {
```

I'm not sure `HttpClientProvider` makes sense as an interface name, it doesn't tell me what the interface does and seems like it would be implemented _by_ an `HTTPClient` and not provide one.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+)
+
+const (
+	defaultMaxRetries              = 3
+	defaultTimeout                 = 1 * time.Second
+	defaultBackoffRetryBaseInMills = 200
+	maxHttpResponseLength          = 5 * 1024 * 1024 // 5MB
+)
+
+type HttpClient struct {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:37]:
```suggestion
type HTTPClient struct {
```

`HTTP` is an initialism and should have consistent capitalization.  This should be addressed in all identifiers that include `HTTP` or `http`.  It might also make sense to simply call this struct `Client` as it will be referenced externally as `httpclient.HTTPClient` or `httpclient.Client`.  Avoiding stutter in scoped names is advised.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	maxRetries              int
+	backoffRetryBaseInMills int
+	client                  clientProvider
+}
+
+type HttpClientProvider interface {
+	Request(path string, ctx context.Context, logger *zap.Logger) ([]byte, error)
+}
+
+type clientProvider interface {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:47]:
```suggestion
type doer interface {
```

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+func New() HttpClientProvider {
+
+	client := &http.Client{Timeout: defaultTimeout}
+
+	return NewHttp(defaultMaxRetries, defaultBackoffRetryBaseInMills, client)
+
+}
+
+func NewHttp(maxRetries int, backoffRetryBaseInMills int, provider clientProvider) HttpClientProvider {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:59]: Can this be folded into the `New()` method through the use of [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis)?  I'd think this might be a better interface:

```go
client := httpclient.New(WithMaxRetries(5), WithClient(&http.Client{TLSConfig: ...}))
```

For the retry base time, use `time.Duration` rather than `int`, that will also allow you to drop the unit statement from the rather long name.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		client:                  provider,
+	}
+
+	return httpClient
+}
+
+func (h *HttpClient) backoffSleep(currentRetryCount int) {
+	backoffInMillis := int64(float64(h.backoffRetryBaseInMills) * math.Pow(2, float64(currentRetryCount)))
+	sleepDuration := time.Millisecond * time.Duration(backoffInMillis)
+	if sleepDuration > 60*1000 {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:72]: Should the maximum `sleepDuration` be configuration exposed to the user?  If they're given control over retries and backoff timing, why not let them specify how long they're willing to wait?

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		sleepDuration = 60 * 1000
+	}
+	time.Sleep(sleepDuration)
+}
+
+func (h *HttpClient) Request(endpoint string, ctx context.Context, logger *zap.Logger) (body []byte, err error) {
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(endpoint, ctx)
+		if err != nil {
+			log.Printf("W! retry [%d/%d], unable to get http response from %s, error: %v", i, h.maxRetries, endpoint, err)
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:82]: Why does this method take a `logger` parameter and then use the stdlib `log.Printf()`?  I also wonder whether the logger needs to be provided with each invocation of `Request()` or can it be included in the `HTTPClient` struct and provided once with the invocation of `New()`.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+			h.backoffSleep(i)
+		}
+	}
+	return
+}
+
+func (h *HttpClient) request(endpoint string, ctx context.Context) ([]byte, error) {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:89]:
```suggestion
func (h *HttpClient) request(ctx context.Context, endpoint string) ([]byte, error) {
```

`ctx` must be the first parameter.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+		return nil, fmt.Errorf("unable to read response body from %s, error: %v", endpoint, err)
+	}
+
+	if len(body) == maxHttpResponseLength {
+		return nil, fmt.Errorf("response from %s, execeeds the maximum length: %v", endpoint, maxHttpResponseLength)
+	}
+	return body, nil
+}
+
+func (h *HttpClient) clientGet(url string, ctx context.Context) (resp *http.Response, err error) {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:125]:
```suggestion
func (h *HttpClient) clientGet(ctx context.Context, url string) (resp *http.Response, err error) {
```

`ctx` must be the first parameter.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	if sleepDuration > 60*1000 {
+		sleepDuration = 60 * 1000
+	}
+	time.Sleep(sleepDuration)
+}
+
+func (h *HttpClient) Request(endpoint string, ctx context.Context, logger *zap.Logger) (body []byte, err error) {
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(endpoint, ctx)
+		if err != nil {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:81]: Not all errors should be retried.  This needs to have some logic for determining which errors are safe to retry and which will never succeed and should result in immediate termination.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to get response from %s, error: %v", endpoint, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unable to get response from %s, status code: %d", endpoint, resp.StatusCode)
+	}
+
+	if resp.ContentLength >= maxHttpResponseLength {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:100]: Why is this check here?  Hasn't the request already been made and the response received.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	} else {
+		reader = resp.Body
+	}
+
+	body, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body from %s, error: %v", endpoint, err)
+	}
+
+	if len(body) == maxHttpResponseLength {
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:119]: Is this necessarily the case?  What about a response that is exactly `maxHTTPResponseLength` bytes?  Does the `io.LimitReader` not provide an error when the limit is reached or some other mechanism?

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
This PR is too large to review and needs to be decomposed.  Let's start with the HTTP client that I've already provided feedback on.  Though I would also say that HTTP clients are well-trod ground and wonder whether there is an existing implementation with retry logic that can be re-used here rather than creating a new implementation.

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 1 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:44]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 2 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:43]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 3 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:37]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 4 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:47]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 5 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:59]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 8 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:89]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 9 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:125]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 7 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:82]: Taskinfo and instanceInfo will just call `Request(endpoints) ` of the HTTPClient. And the `httpclient` will be initialize once to use for both of them. So I think it's better to put the logger in the `Request()`
And I will change all the` log.Printf()` to use logger.
Thanks!

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 6 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:72]: All these parameters will not be changed through the config file, so they will not be controlled by customer.

**Comment by JohnWu20** [comment]
> Hi @JohnWu20 please fix the failing tests listed below.
> 
> ```
> windows-test - Failed
> integration-tests - Failed
> unit-tests - Failed
> lint - Failed
> ```

Thanks Alolita, I have fixed the failure, and currently the 
loadtest - Failed which I believe there is no relation-ship with my code change.
@Aneurysm9  I have split the  to two parts, and here is anther [pr](url)  Please help me review it

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 11 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go]: This httpclient will mainly use for ECS Introspection endpoint. And sometimes, when a task have too many containers, this response will be very large that we don't want to deal with this situation.

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 12 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go]: This part of logic just wants to keep same with the situation that we know the length. It means that we don't want to deal with the response when its body is too big.

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+		cii.logger.Warn("Failed to get instance id from arn, error: ", zap.Error(err))
+	}
+
+	cii.Lock()
+	cii.clusterName = cluster
+	cii.containerInstanceID = instanceID
+	cii.Unlock()
+
+	if cii.clusterName != "" && cii.containerInstanceID != "" {
+		close(cii.readyC)
```
Comment (Aneurysm9) [inline receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:110]: What ensures this isn't closed twice?  Is there any way to eliminate this channel?  It seems to only be used for tests, which is a nasty code smell.

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	ecsTaskEndpointProvider hostIPProvider
+	runningTaskCount        int64
+	runningTasksInfo        []ECSTask
+	readyC                  chan bool
+	sync.RWMutex
+}
+
+type taskInfoOption func(*taskInfo)
+
+func newECSTaskInfo(ctx context.Context, ecsTaskEndpointProvider hostIPProvider,
```
Comment (Aneurysm9) [inline receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go:61]: All of the same comments from the instance info file seem to apply here as well.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	defaultMaxRetries              = 3
+	defaultTimeout                 = 1 * time.Second
+	defaultBackoffRetryBaseInMills = 200
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:32]: Thanks will change it

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+	maxHTTPResponseLength          = 5 * 1024 * 1024 // 5MB
+)
+
+type HTTPClient struct {
+	maxRetries              int
+	backoffRetryBaseInMills time.Duration
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:38]: Thanks will change it

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+	for _, opt := range options {
+		opt(httpClient)
+	}
+
+	return httpClient
+}
+
+func (h *HTTPClient) backoffSleep(currentRetryCount int) {
+	backoffInMillis := int64(float64(h.backoffRetryBaseInMills) * math.Pow(2, float64(currentRetryCount)))
+	sleepDuration := time.Millisecond * time.Duration(backoffInMillis)
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:75]: Thanks, will change it.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	time.Sleep(sleepDuration)
+}
+
+func (h *HTTPClient) Request(ctx context.Context, endpoint string, logger *zap.Logger) (body []byte, err error) {
+	for i := 0; i < h.maxRetries; i++ {
+		body, err = h.request(ctx, endpoint)
+		if err != nil {
+			sugar := logger.Sugar()
+			sugar.Warn("W! retry [%d/%d], unable to get http response from %s, error: %v", i, h.maxRetries, endpoint, err)
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:87]: Thanks, will change it

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+		readyC:                   readyC,
+	}
+
+	for _, opt := range options {
+		opt(cii)
+	}
+
+	shouldRefresh := func() bool {
+		//stop the refresh once we get instance ID and type successfully
+		return cii.clusterName == "" || cii.containerInstanceID == ""
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:76]: Thanks, will change it

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go (hunk 13 above)
Comment (JohnWu20) [reply receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:110]: Will add logic to check if the channel is closed or not

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+func (cii *containerInstanceInfo) GetClusterName() string {
+	cii.Lock()
+	defer cii.Unlock()
+	return cii.clusterName
+}
+
+func (cii *containerInstanceInfo) GetContainerInstanceID() string {
+	cii.Lock()
+	defer cii.Unlock()
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:122]: Will change it, thanks

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+	cii.Unlock()
+
+	if cii.clusterName != "" && cii.containerInstanceID != "" {
+		close(cii.readyC)
+	}
+}
+
+func (cii *containerInstanceInfo) GetClusterName() string {
+	cii.Lock()
+	defer cii.Unlock()
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:116]: Will change it, thanks

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+
+type ContainerInstance struct {
+	Cluster              string
+	ContainerInstanceArn string
+}
+
+type ecsInstanceInfoOption func(*containerInstanceInfo)
+
+func newECSInstanceInfo(ctx context.Context, ecsAgentEndpointProvider hostIPProvider,
+	refreshInterval time.Duration, logger *zap.Logger, httpClient Requester, readyC chan bool, options ...ecsInstanceInfoOption) containerInstanceInfoProvider {
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:61]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient_test.go [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+type fakeClient struct {
+	response *http.Response
+	err      error
+}
+
+func (f *fakeClient) Do(req *http.Request) (*http.Response, error) {
+	return f.response, nil
+}
+
+func TestRequestSecuess(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient_test.go:37]: Will change it, thanks

File: internal/aws/containerinsight/httpclient/httpclient.go (hunk 10 above)
Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:85]: Changed it using new retryablehttp package.

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+
+	if len(body) == maxHTTPResponseLength {
+		return nil, fmt.Errorf("response from %s, execeeds the maximum length: %v", endpoint, maxHTTPResponseLength)
+	}
+	return body, nil
+
+}
+
+func (h *HTTPClient) clientGet(ctx context.Context, url string) (resp *http.Response, err error) {
+	req, err := retryablehttp.NewRequest("GET", url, nil)
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient.go:112]:
```suggestion
	req, err := retryablehttp.NewRequest(http.MethodGet, url, nil)
```

File: internal/aws/containerinsight/httpclient/httpclient_test.go [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+
+	assert.NotNil(t, body)
+
+}
+
+func TestRequestWithFailedStatus(t *testing.T) {
+
+	respBody := "body"
+	response := &http.Response{
+		Status:        "200 OK",
```
Comment (Aneurysm9) [inline internal/aws/containerinsight/httpclient/httpclient_test.go:95]:
```suggestion
		Status:        "Bad Request",
```

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+	cii.logger.Info("Fetch instance id and type from ec2 metadata")
+	resp, err := cii.httpClient.Request(ctx, cii.getECSAgentEndpoint())
+	if err != nil {
+		cii.logger.Warn("Failed to call ecsagent endpoint, error: ", zap.Error(err))
+	}
+
+	err = json.Unmarshal(resp, containerInstance)
+	if err != nil {
+		cii.logger.Warn("Failed to unmarshal: ", zap.Error(err))
+		cii.logger.Warn("Resp content is " + string(resp))
```
Comment (Aneurysm9) [inline receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go:89]:
```suggestion
		cii.logger.Debug("Resp content is " + string(resp))
```

This may be a bit much to log, up to 5MB.

File: receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+	resp, err := ti.httpClient.Request(ctx, ti.getECSAgentTaskInfoEndpoint())
+	if err != nil {
+		ti.logger.Warn("Failed to call ecsagent taskinfo endpoint, error: ", zap.Error(err))
+		return ecsTasksInfo
+	}
+
+	err = json.Unmarshal(resp, ecsTasksInfo)
+	if err != nil {
+		ti.logger.Warn("Unable to parse resp from ecsagent taskinfo endpoint, error:", zap.Error(err))
+		ti.logger.Warn("D! resp content is %s" + string(resp))
```
Comment (Aneurysm9) [inline receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_task_info.go:88]:
```suggestion
		ti.logger.Debug("D! resp content is %s" + string(resp))
```

**Review by Aneurysm9 — APPROVED** [review]

File: internal/aws/containerinsight/httpclient/httpclient.go [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+type HTTPClient struct {
+	client doer
+}
+
+type Requester interface {
+	Request(ctx context.Context, path string) ([]byte, error)
+}
+
+type doer interface {
+	Do(request *retryablehttp.Request) (*http.Response, error)
```
Comment (bogdandrutu) [inline internal/aws/containerinsight/httpclient/httpclient.go:45]: We do have already a queued-retry logic for exporters, why is this needed?

Comment (JohnWu20) [reply internal/aws/containerinsight/httpclient/httpclient.go:45]: This is used for calling the ECS container agent introspection task endpoint and cluster endpoint, and there are some specific setting for the response data, I think this is different from the logic for exporters.

Comment (bogdandrutu) [reply internal/aws/containerinsight/httpclient/httpclient.go:45]: @Aneurysm9 you know both, please followup if this is a duplicate logic to ensure we use the exporter helper if possible :)
